commit 9c3d87366941de3099fefd741d7ba4a5d3785bba
Author:     Yuhong Wen <yuhongw@nvidia.com>
AuthorDate: Fri Feb 17 10:44:52 2023 -0500
Commit:     GitHub <noreply@github.com>
CommitDate: Fri Feb 17 10:44:52 2023 -0500

     FL Client integrate with FCI Cellnet (#1380)
    
    * FL server integrate with FCI cellnet.
    
    * Codestyle reformat.
    
    * fed_server_test.py integrate with Cellnet.
    
    * FL client integrate with Cellnet. Client register to server.
    
    * Reformat codestyle.
    
    * reformat codestyle.
    
    * update the message.
    
    * Codestyle reformat.
    
    * Fixed the import sort.
    
    * Addes the PR reviews.
    
    * codestyle fix.
    
    * made the cell_timeout configurable.
    
    * codestyle fix.
    
    * removed no use import.
    
    * disable simulator_runner_test temporary.
---
 nvflare/private/fed/client/client_engine.py   |  25 ++--
 nvflare/private/fed/client/client_executor.py | 180 +++++++++++---------------
 2 files changed, 83 insertions(+), 122 deletions(-)

diff --git a/nvflare/private/fed/client/client_engine.py b/nvflare/private/fed/client/client_engine.py
index 61071b47..dced5698 100644
--- a/nvflare/private/fed/client/client_engine.py
+++ b/nvflare/private/fed/client/client_engine.py
@@ -18,13 +18,11 @@ import re
 import shutil
 import sys
 import threading
-import time
 
 from nvflare.apis.event_type import EventType
 from nvflare.apis.fl_component import FLComponent
 from nvflare.apis.fl_constant import MachineStatus, WorkspaceConstants
 from nvflare.apis.fl_context import FLContext, FLContextManager
-from nvflare.apis.shareable import Shareable
 from nvflare.apis.workspace import Workspace
 from nvflare.fuel.utils.network_utils import get_open_ports
 from nvflare.private.defs import ERROR_MSG_PREFIX, ClientStatusKey, EngineConstant
@@ -49,7 +47,7 @@ def _remove_custom_path():
 class ClientEngine(ClientEngineInternalSpec):
     """ClientEngine runs in the client parent process."""
 
-    def __init__(self, client, client_name, sender, args, rank, workers=5):
+    def __init__(self, client, client_name, args, rank, workers=5):
         """To init the ClientEngine.
 
         Args:
@@ -63,11 +61,9 @@ class ClientEngine(ClientEngineInternalSpec):
         super().__init__()
         self.client = client
         self.client_name = client_name
-        self.sender = sender
         self.args = args
         self.rank = rank
-        self.client.process = None
-        self.client_executor = ProcessExecutor(os.path.join(args.workspace, "startup"))
+        self.client_executor = ProcessExecutor(client, os.path.join(args.workspace, "startup"))
         self.admin_agent = None
 
         self.fl_ctx_mgr = FLContextManager(
@@ -212,7 +208,7 @@ class ClientEngine(ClientEngineInternalSpec):
         touch_file = os.path.join(self.args.workspace, "shutdown.fl")
         self.fire_event(EventType.SYSTEM_END, self.new_context())
 
-        thread = threading.Thread(target=_shutdown_client, args=(self.client, self.admin_agent, touch_file))
+        thread = threading.Thread(target=shutdown_client, args=(self.client, touch_file))
         thread.start()
 
         return "Shutdown the client..."
@@ -221,7 +217,7 @@ class ClientEngine(ClientEngineInternalSpec):
         self.logger.info("Client shutdown...")
         touch_file = os.path.join(self.args.workspace, "restart.fl")
         self.fire_event(EventType.SYSTEM_END, self.new_context())
-        thread = threading.Thread(target=_shutdown_client, args=(self.client, self.admin_agent, touch_file))
+        thread = threading.Thread(target=shutdown_client, args=(self.client, touch_file))
         thread.start()
 
         return "Restart the client..."
@@ -253,27 +249,22 @@ class ClientEngine(ClientEngineInternalSpec):
     def reset_errors(self, job_id):
         self.client_executor.reset_errors(job_id)
 
-    def send_aux_command(self, shareable: Shareable, job_id):
-        return self.client_executor.process_aux_command(shareable, job_id)
-
     def get_all_job_ids(self):
         return self.client_executor.get_run_processes_keys()
 
 
-def _shutdown_client(federated_client, admin_agent, touch_file):
+def shutdown_client(federated_client, touch_file):
     with open(touch_file, "a"):
         os.utime(touch_file, None)
 
     try:
         print("About to shutdown the client...")
         federated_client.communicator.heartbeat_done = True
-        time.sleep(3)
+        # time.sleep(3)
         federated_client.close()
 
-        if federated_client.process:
-            federated_client.process.terminate()
-
-        admin_agent.shutdown()
+        federated_client.status = ClientStatus.STOPPED
+        # federated_client.cell.stop()
         security_close()
     except BaseException as e:
         secure_log_traceback()
diff --git a/nvflare/private/fed/client/client_executor.py b/nvflare/private/fed/client/client_executor.py
index 6a7dcc97..84dddc1a 100644
--- a/nvflare/private/fed/client/client_executor.py
+++ b/nvflare/private/fed/client/client_executor.py
@@ -20,11 +20,13 @@ import sys
 import threading
 import time
 from abc import ABC, abstractmethod
-from multiprocessing.connection import Client
 
-from nvflare.apis.fl_constant import AdminCommandNames, ReturnCode, RunProcessKey
+from nvflare.apis.fl_constant import AdminCommandNames, RunProcessKey
 from nvflare.apis.resource_manager_spec import ResourceManagerSpec
-from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.fuel.f3.cellnet.cell import FQCN
+from nvflare.fuel.f3.cellnet.defs import MessageHeaderKey, ReturnCode
+from nvflare.fuel.utils import fobs
+from nvflare.private.defs import CellChannel, new_cell_message
 from nvflare.security.logging import secure_format_exception, secure_log_traceback
 
 from .client_status import ClientStatus, get_status_message
@@ -117,46 +119,22 @@ class ClientExecutor(ABC):
             job_id: the job_id
         """
 
-    @abstractmethod
-    def process_aux_command(self, shareable: Shareable, job_id):
-        """Processes the aux command.
-
-        Args:
-            shareable: aux message Shareable
-            job_id: the job_id
-        """
-        pass
-
 
 class ProcessExecutor(ClientExecutor):
     """Run the Client executor in a child process."""
 
-    def __init__(self, startup):
+    def __init__(self, client, startup):
         """To init the ProcessExecutor.
 
         Args:
             startup: startup folder
         """
+        self.client = client
         self.logger = logging.getLogger(self.__class__.__name__)
         self.startup = startup
         self.run_processes = {}
         self.lock = threading.Lock()
 
-    def get_conn_client(self, job_id):
-        # should be call within self.lock
-        listen_port = self.run_processes.get(job_id, {}).get(RunProcessKey.LISTEN_PORT)
-        conn_client = self.run_processes.get(job_id, {}).get(RunProcessKey.CONNECTION, None)
-
-        if not conn_client:
-            try:
-                address = ("localhost", listen_port)
-                conn_client = Client(address, authkey="client process secret password".encode())
-                self.run_processes[job_id][RunProcessKey.CONNECTION] = conn_client
-            except Exception:
-                pass
-
-        return conn_client
-
     def start_app(
         self,
         client,
@@ -203,7 +181,7 @@ class ProcessExecutor(ClientExecutor):
             + " -c "
             + client.client_name
             + " -p "
-            + str(listen_port)
+            + str(client.cell.get_internal_listener_url())
             + " -g "
             + target
             + " -s fed_client.json "
@@ -241,12 +219,18 @@ class ProcessExecutor(ClientExecutor):
         """
         try:
             with self.lock:
-                conn_client = self.get_conn_client(job_id)
-
-                if conn_client:
-                    data = {"command": AdminCommandNames.CHECK_STATUS, "data": {}}
-                    conn_client.send(data)
-                    status_message = conn_client.recv()
+                data = {}
+                fqcn = FQCN.join([self.client.client_name, job_id])
+                request = new_cell_message({}, fobs.dumps(data))
+                return_data = self.client.cell.send_request(
+                    target=fqcn,
+                    channel=CellChannel.CLIENT_COMMAND,
+                    topic=AdminCommandNames.CHECK_STATUS,
+                    request=request,
+                )
+                return_code = return_data.get_header(MessageHeaderKey.RETURN_CODE)
+                if return_code == ReturnCode.OK:
+                    status_message = fobs.loads(return_data.payload)
                     self.logger.debug("check status from process listener......")
                     return status_message
                 else:
@@ -268,12 +252,15 @@ class ProcessExecutor(ClientExecutor):
         """
         try:
             with self.lock:
-                conn_client = self.get_conn_client(job_id)
-
-                if conn_client:
-                    data = {"command": AdminCommandNames.SHOW_STATS, "data": {}}
-                    conn_client.send(data)
-                    run_info = conn_client.recv()
+                data = {}
+                fqcn = FQCN.join([self.client.client_name, job_id])
+                request = new_cell_message({}, fobs.dumps(data))
+                return_data = self.client.cell.send_request(
+                    target=fqcn, channel=CellChannel.CLIENT_COMMAND, topic=AdminCommandNames.SHOW_STATS, request=request
+                )
+                return_code = return_data.get_header(MessageHeaderKey.RETURN_CODE)
+                if return_code == ReturnCode.OK:
+                    run_info = fobs.loads(return_data.payload)
                     return run_info
                 else:
                     return {}
@@ -293,12 +280,18 @@ class ProcessExecutor(ClientExecutor):
         """
         try:
             with self.lock:
-                conn_client = self.get_conn_client(job_id)
-
-                if conn_client:
-                    data = {"command": AdminCommandNames.SHOW_ERRORS, "data": {}}
-                    conn_client.send(data)
-                    errors_info = conn_client.recv()
+                data = {"command": AdminCommandNames.SHOW_ERRORS, "data": {}}
+                fqcn = FQCN.join([self.client.client_name, job_id])
+                request = new_cell_message({}, fobs.dumps(data))
+                return_data = self.client.cell.send_request(
+                    target=fqcn,
+                    channel=CellChannel.CLIENT_COMMAND,
+                    topic=AdminCommandNames.SHOW_ERRORS,
+                    request=request,
+                )
+                return_code = return_data.get_header(MessageHeaderKey.RETURN_CODE)
+                if return_code == ReturnCode.OK:
+                    errors_info = return_data.payload
                     return errors_info
                 else:
                     return None
@@ -315,35 +308,20 @@ class ProcessExecutor(ClientExecutor):
         """
         try:
             with self.lock:
-                conn_client = self.get_conn_client(job_id)
+                data = {"command": AdminCommandNames.RESET_ERRORS, "data": {}}
+                fqcn = FQCN.join([self.client.client_name, job_id])
+                request = new_cell_message({}, fobs.dumps(data))
+                self.client.cell.fire_and_forget(
+                    targets=fqcn,
+                    channel=CellChannel.CLIENT_COMMAND,
+                    topic=AdminCommandNames.RESET_ERRORS,
+                    message=request,
+                )
 
-                if conn_client:
-                    data = {"command": AdminCommandNames.RESET_ERRORS, "data": {}}
-                    conn_client.send(data)
         except Exception as e:
             self.logger.error(f"reset_errors execution exception: {secure_format_exception(e)}.")
             secure_log_traceback()
 
-    def process_aux_command(self, shareable: Shareable, job_id):
-        """Processes the aux command.
-
-        Args:
-            shareable: aux message Shareable
-            job_id: the job_id
-        """
-        try:
-            with self.lock:
-                conn_client = self.get_conn_client(job_id)
-                if conn_client:
-                    data = {"command": AdminCommandNames.AUX_COMMAND, "data": shareable}
-                    conn_client.send(data)
-                    reply = conn_client.recv()
-                    return reply
-                else:
-                    return make_reply(ReturnCode.EXECUTION_EXCEPTION)
-        except Exception:
-            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
-
     def abort_app(self, job_id):
         """Aborts the running app.
 
@@ -359,13 +337,19 @@ class ProcessExecutor(ClientExecutor):
                 if process_status == ClientStatus.STARTED:
                     try:
                         child_process = self.run_processes[job_id][RunProcessKey.CHILD_PROCESS]
-                        conn_client = self.get_conn_client(job_id)
-                        if conn_client:
-                            data = {"command": AdminCommandNames.ABORT, "data": {}}
-                            conn_client.send(data)
-                            self.logger.debug("abort sent")
-
-                        threading.Thread(target=self._terminate_process, args=[child_process, job_id]).start()
+                        data = {}
+                        fqcn = FQCN.join([self.client.client_name, job_id])
+                        request = new_cell_message({}, fobs.dumps(data))
+                        return_data = self.client.cell.fire_and_forget(
+                            targets=fqcn,
+                            channel=CellChannel.CLIENT_COMMAND,
+                            topic=AdminCommandNames.ABORT,
+                            message=request,
+                        )
+                        self.logger.debug("abort sent")
+
+                        if child_process:
+                            threading.Thread(target=self._terminate_process, args=[child_process, job_id]).start()
                         self.run_processes.pop(job_id)
                         break
                     except Exception as e:
@@ -376,9 +360,6 @@ class ProcessExecutor(ClientExecutor):
                             secure_log_traceback()
                         retry -= 1
                         time.sleep(5.0)
-                    finally:
-                        if conn_client:
-                            conn_client.close()
                 else:
                     self.logger.info(f"Client worker process for run: {job_id} was already terminated.")
                     break
@@ -406,33 +387,25 @@ class ProcessExecutor(ClientExecutor):
         with self.lock:
             process_status = self.run_processes.get(job_id, {}).get(RunProcessKey.STATUS, ClientStatus.NOT_STARTED)
             if process_status == ClientStatus.STARTED:
-                conn_client = self.get_conn_client(job_id)
-                if conn_client:
-                    data = {"command": AdminCommandNames.ABORT_TASK, "data": {}}
-                    conn_client.send(data)
-                    self.logger.debug("abort_task sent")
+                data = {"command": AdminCommandNames.ABORT_TASK, "data": {}}
+                fqcn = FQCN.join([self.client.client_name, job_id])
+                request = new_cell_message({}, fobs.dumps(data))
+                return_data = self.client.cell.fire_and_forget(
+                    targets=fqcn,
+                    channel=CellChannel.CLIENT_COMMAND,
+                    topic=AdminCommandNames.ABORT_TASK,
+                    message=request,
+                )
+                self.logger.debug("abort_task sent")
 
     def _wait_child_process_finish(self, client, job_id, allocated_resource, token, resource_manager):
-        # wait for the listen_command thread to start, and send "start" message to wake up the connection.
-        start = time.time()
-        while True:
-            with self.lock:
-                conn_client = self.get_conn_client(job_id)
-                if conn_client:
-                    data = {"command": AdminCommandNames.START_APP, "data": {}}
-                    conn_client.send(data)
-                    break
-            time.sleep(1.0)
-            if time.time() - start > 15:
-                break
-
         self.logger.info(f"run ({job_id}): waiting for child worker process to finish.")
         with self.lock:
             child_process = self.run_processes.get(job_id, {}).get(RunProcessKey.CHILD_PROCESS)
         if child_process:
             child_process.wait()
-            return_code = child_process.returncode
-            self.logger.info(f"run ({job_id}): child worker process finished with execution code: {return_code}")
+            # return_code = child_process.returncode
+            self.logger.info(f"run ({job_id}): child worker process finished.")
 
         if allocated_resource:
             resource_manager.free_resources(
@@ -440,9 +413,6 @@ class ProcessExecutor(ClientExecutor):
             )
 
         with self.lock:
-            conn_client = self.get_conn_client(job_id)
-            if conn_client:
-                conn_client.close()
             if job_id in self.run_processes.keys():
                 self.run_processes.pop(job_id)
 
